content/blog/2022-05-23-vectorization/index.md

```diff
@@ -0,0 +1,238 @@
+
+Next, loops in the code are fully unrolled, if possible. Unrolling only applies to loops that have one iteration variable that is increasing by 1 or decreasing by 1. Further, the loop header must have exactly one branch condition, depending on the iteration variable. The loop also must have only one exit, from the branch condition in the header. By unrolling the code, loops get expanded into long sequences of straight-line code without branches or jumps. Having straight line code allows for long sequences of instructions that can be identified as vectorizable, and the possibility that a vector may be used several times.
+
+After unrolling, stores are then moved as late in the basic block as possible. Store movement depends on alias analysis, because a store can only move past another store or another load, assuming no aliasing. Moving stores later in the basic block allows for a greater opportunity of finding vectors, because vectors have to end when a store is identified. In the same vein, `const` instructions and `id` instructions are moved as early as possible, while `print` instructions are moved to be as far down in a basic block as possible.
+
+Following moving stores, each basic block is made canonical in form such that sequences of adds, divides, multiplies, and subtractions occur consecutively as much as possible. This pass is used to identify potential vectors in a simple manner.
+
+Finally, block coalescing is done to the control flow graph. When loop unrolling is complete, there may exist extraneous jumps in the control flow graph. Basic blocks connected by a jump are stitched together into a single block, assuming that the original target label of the jump is unused after the jump is eliminated. By coalescing the control flow graph, the vectorization algorithms are able to search longer jump-free sequences of instructions and possibly find more vectors than had the coalescing not been applied.
+
+### Naive Vectorization
+
+In naive vectorization, given a run of instructions, a vector will be created, with loads and stores created appropriately. There is no attempt to reuse a created vector. As an example of naive vectorization, given the run: 
+
+```bril
+w: int = add a b;
+x: int = add b c;
+y: int = add a c;
+z: int = add d d;
+```
+
+then the below instructions are created, and used to replace the original run of instructions:
+
+```bril
+v1: vector = veczero;
+v2: vector = veczero;
+
+one: int = const 1;
+
+i: int = const 0;
+vecload v1 i a;
+i: int = add i one;
+vecload v1 i b;
+i: int = add i one;
+vecload v1 i a;
+i: int = add i one;
+vecload v1 i d;
+
+i: int = const 0;
+vecload v2 i b;
+i: int = add i one;
+vecload v2 i c;
+i: int = add i one;
+vecload v2 i c;
+i: int = add i one;
+vecload v2 i d;
+
+sum: vector = vecadd v1 v2;
+
+i: int = const 0;
+w: int = vecstore sum i;
+i: int = add i one;
+x: int = vecstore sum i;
+i: int = add i one;
+y: int = vecstore sum i;
+i: int = add i one;
+z: int = vecstore sum i;
+```
```

Comment:
Very cool! Looking at this code makes me wonder whether *downstream* analyses (e.g., "opportunistic vectorization") wouldn't have been a little easier if there were a single instruction pack/unpack entire vectors from/to scalar registers…

Reply:
A single unpack would have been very useful in this circumstance, at least from the view of writing to code to generate all the `vecstore` operations.

Reply:
Yeah, that definitely makes sense for this version. Bril doesn't really have a way to write to multiple registers at once anyway, so making this work would be pretty hard.